File: src/main/java/com/github/urlshorten/UrlShortenService.java
package com.github.urlshorten;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.net.MalformedURLException;
import java.net.URL;
import java.time.LocalDateTime;
import java.util.Objects;
import java.util.Optional;

@Service
public class UrlShortenService {

    private final UrlRepository urlRepository;

    public UrlShortenService(UrlRepository urlRepository) {
        this.urlRepository = urlRepository;
    }

    @Transactional
    public Url shorten(String url, IdStrategy strategy) {
        Objects.requireNonNull(url);
        Objects.requireNonNull(strategy);

        Url urlForShorten = new Url();
        urlForShorten.setCreationDate(LocalDateTime.now());
        urlForShorten.setExpirationDate(urlForShorten.getCreationDate().plusYears(2L));
        urlForShorten.setOriginalUrl(url);

        urlRepository.save(urlForShorten);

        String id = strategy.apply(urlForShorten.getId());
        try {
            URL netUrl = new URL(url);

            urlForShorten.setShortenUrl(url.replace(netUrl.getPath(), id));
        } catch (MalformedURLException e) {
            throw new IllegalArgumentException(e);
        }

        return urlForShorten;
    }

    @Transactional(readOnly = true)
    public Optional<Url> retrieve(String shortenUrl) {
        Objects.requireNonNull(shortenUrl);

        return urlRepository.findByShortenUrl(shortenUrl);
    }
}

File: src/main/kotlin/com/github/urlshorten/UrlShortenService.kt
package com.github.urlshorten

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.net.MalformedURLException
import java.net.URL
import java.time.LocalDateTime

@Service
class UrlShortenService(val urlRepository: UrlRepository) {

    @Transactional
    open fun shorten(url: String, strategy: IdStrategy): Url? {
        var urlForShorten = Url(
                id = null,
                creationDate = LocalDateTime.now(),
                expirationDate = LocalDateTime.now().plusYears(2L),
                originalUrl = url,
                shortenUrl = null,
        )
        urlForShorten = urlRepository.save(urlForShorten)
        val id = strategy.apply(urlForShorten.id!!)
        try {
            val netUrl = URL(url)
            urlForShorten.shortenUrl = url.replace(netUrl.path, "/$id")
        } catch (e: MalformedURLException) {
            throw IllegalArgumentException(e)
        }
        return urlForShorten
    }

    @Transactional(readOnly = true)
    open fun retrieve(shortenUrl: String): Url? = urlRepository.findByShortenUrl(shortenUrl)
}